plugins/outputs/mqtt/topic_name_generator.go

```diff
@@ -0,0 +1,51 @@
+package mqtt
+
+import (
+	"strings"
+	"text/template"
+
+	"github.com/influxdata/telegraf"
+)
+
+type TopicNameGenerator struct {
+	Hostname    string
+	TopicPrefix string
+	metric      telegraf.Metric
+	template    *template.Template
+}
+
+func NewTopicNameGenerator(topicPrefix string, temp *template.Template) *TopicNameGenerator {
+	return &TopicNameGenerator{TopicPrefix: topicPrefix, template: temp}
+}
+
+func (t *TopicNameGenerator) Tag(key string) string {
+	tagString, _ := t.metric.GetTag(key)
+	return tagString
+}
+
+// PluginName returns the name of the plugin and is used by the template parser when the topic uses {{ .PluginName }}
+func (t *TopicNameGenerator) PluginName() string {
+	return t.metric.Name()
+}
+
+func (t *TopicNameGenerator) Generate(hostname string, m telegraf.Metric) (string, error) {
+	t.Hostname = hostname
+	t.metric = m
+	var b strings.Builder
+	err := t.template.Execute(&b, t)
+	if err != nil {
+		return "", err
+	}
+	var ts []string
+	for _, p := range strings.Split(b.String(), "/") {
+		if p != "" {
+			ts = append(ts, p)
+		}
+	}
+	topic := strings.Join(ts, "/")
+	// This is to keep backward compatibility with previous behaviour where the plugin name was always present
+	if topic == "" {
+		return m.Name(), nil
```

Comment:
@srebhan I have done as you recommend here, however, I can't get rid of the `PluginName` if I want to leave it as part of the keywords used to create a topic name.

I may be just lacking knowledge on the template library though, so maybe you can point me in a different direction if that's the case, but I believe, I have to leave in place that function. What I have done though is to add a comment to the function to explain _why_ is relevant to keep it around.

Reply:
But does it have to be a function to access it? Why not make it a struct field?

Reply:
No, it didn't need to be a function. I have updated to be a struct property instead 👍